client/src/components/ResourceList/index.js

```diff
@@ -1,29 +1,101 @@
 import PropTypes from "prop-types";
+import { useEffect, useState } from "react";
 
+import { FormControls } from "../../components";
+import { TopicService, useService, ResourceService } from "../../services";
 import "./ResourceList.scss";
 
-export default function ResourceList({ publish, resources }) {
+export default function ResourceList({ publish, resources, pathname }) {
+	const [topics, setTopics] = useState([]);
+	const topicService = useService(TopicService);
+	const [selectedTopic, setSelectedTopic] = useState("");
+	const [filteredResources, setFilteredResources] = useState([]);
+	const resourceService = useService(ResourceService);
+
+	useEffect(() => {
+		const fetchTopics = async () => {
+			try {
+				const isTestEnvironment = process.env.NODE_ENV === "test";
+
+				const fetchedTopics = isTestEnvironment
+					? []
+					: await topicService.getTopics();
+
+				setTopics(fetchedTopics);
+			} catch (error) {
+				throw new Error(`Error fetching topics: ${error.message}`);
+			}
+		};
+
+		fetchTopics();
+	}, [setTopics, topicService]);
+
+	useEffect(() => {
+		const fetchResourcesByTopic = async () => {
+			try {
+				const allResources = await resourceService.getPublished({});
+				const filtered = allResources.resources.filter(
+					(topic) => topic.topic_name === selectedTopic
+				);
+				setFilteredResources(filtered);
+			} catch (error) {
+				throw new Error(`Error fetching resources: ${error.message}`);
+			}
+		};
+
+		if (selectedTopic) {
+			fetchResourcesByTopic();
+		} else {
+			setFilteredResources(resources);
+		}
+	}, [selectedTopic, resourceService, resources]);
+
+	const handleChange = (event) => {
+		const selectedValue = event.target.value;
+		const selectedOption = topics.find((option) => option.id === selectedValue);
+		setSelectedTopic(selectedOption ? selectedOption.name : "");
+	};
+
 	return (
-		<ul className="resource-list">
-			{resources.length === 0 && (
-				<li className="no-resources">
-					<em>No resources to show.</em>
-				</li>
+		<>
+			{resources.length > 0 && pathname !== "/drafts" && (
```

Comment:
❓ I'm in two minds about the check on `resources.length` here. On the one hand yes there's no point letting the user filter no resources, but on the other hand we will happily show no resources for a given topic _after_ they've filtered the list! Might be simpler and more consistent to leave it out.

❌ I really don't like the `pathname` handling here. `!== "/drafts"` (rather than `=== "/"`) means that every time we add a new display of resources we have to explicitly opt _out_ of this behaviour, instead of opting _in_ when we actually wanted it.

💡 But if this is basically just a boolean toggle, "allow topic filter if `"/"`, don't if `"/drafts"`", why not simply let the parent _pass a boolean_, e.g. `allowTopicFiltering`? That also means we don't have to maintain a growing array e.g. `["/", "/foo", "/bar"].includes(pathname)` as we add more uses, which breaks the [open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

Reply:
> ❓ I'm in two minds about the check on `resources.length` here. On the one hand yes there's no point letting the user filter no resources, but on the other hand we will happily show no resources for a given topic _after_ they've filtered the list! Might be simpler and more consistent to leave it out.
> 
> ❌ I really don't like the `pathname` handling here. `!== "/drafts"` (rather than `=== "/"`) means that every time we add a new display of resources we have to explicitly opt _out_ of this behaviour, instead of opting _in_ when we actually wanted it.
> 
> 💡 But if this is basically just a boolean toggle, "allow topic filter if `"/"`, don't if `"/drafts"`", why not simply let the parent _pass a boolean_, e.g. `allowTopicFiltering`? That also means we don't have to maintain a growing array e.g. `["/", "/foo", "/bar"].includes(pathname)` as we add more uses, which breaks the [open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

@textbook that's done